docs/designs/ruby_test_plan_002.md

```diff
@@ -0,0 +1,150 @@
+Feature: Cachi2 can prefetch dependencies for Ruby projects
+    
+    Scenario outline: a user wants to fetch dependencies for a package containing well-defined Gemfile.lock
+       Given  clean repository of a "well-defined project"
+        When  a user request to fetch dependencies for a "well-defined project"
+        Then  direct and transitive dependencies are fetched
+         And  dependencies could be found within the "well-defined project"
+    Examples:
+        |    well-defined project                          |
+        | Depends on gems from rubygems.org only           |
+        | Has correct .gemspec                             |
+        | Has correct Git dependency                       |
+        | Has several correct Git dependencies             |
+        | Has path dependency                              |
+        | Has several path dependencies                    |
+        | Has a mixture of gems, Git and path dependencies |
+
+    Scenario outline: a user wants to fetch dependencies for a package containing ill-defined Gemfile.lock
+       Given  clean repository of a "ill-defined project"
+        When  a user request to fetch dependencies for a "ill-defined project"
+        Then  Cachi2 reports an error
+         And  repository for "ill-defined project" remains unchanged
+    Examples:
+        |    ill-defined project                     |
+        | Gemfile.lock is missing                    |
+        | Gemfile.lock contains critical code        |
+        | Git dependency is not pinned to hash       |
+        | Gem dependency is sourced from unknown URL |
+        | Transitive dependency is ill-defined       |
+    Note: "unknown URL" is any URL other than https://rubygems.org
+
+    Scenario outline: a user wants to fetch dependencies for a package containing poorly-defined Gemfile.lock
+        When  a user requests to fetch dependnecies for a "poorly-defined project"
+        Then  direct and transitive dependencies are fetched
+         And  could be found within the "poorly-defined project"
+         But  Cachi2 warns about detected problem
+    Examples:
+        |    poorly-defined project             |
+        | Gemfile contains executable code      |
+TODO: Should we detect, warn and block packages with any code in Gemfile?
+      Do we even want to try to recover? If not then this scenario is not needed.
+
+
+@integration_test
+Feature: Cachi2 can preset configuration for hermetic builds
+
+    Scenario outline: a user prepares a repository with single Ruby project
+        When  a user generates environment for "single Ruby project"
+        Then  the environment contains a pointer to local Bundler config
+         And  Bundler cache directory structure is recreated for every project
+    Examples:
+        |    single Ruby project                            |
+        | Ruby project with just gem dependencies           |
+        | Ruby project with gem and git dependencies        |
```

Comment:
Providing the build configuration is essentially setting some configuration keys in `.bundle/config` (or setting env vars), so I don't think the type of dependencies affect the ouput here.

Reply:
We can probably have two scenarios for preset configuration: one for a single project, and one for a multi-project (or maybe even only having the multi-project scenario by itself is good enough, considering we also have the e2e that covers this)

Reply:
The first approach sounds fine-grained, I'll use it.